bin/faucet/src/faucet/updates.rs

```diff
@@ -0,0 +1,114 @@
+use std::convert::Infallible;
+
+use axum::response::sse::Event;
+use base64::{Engine, engine::general_purpose};
+use miden_objects::{
+    account::AccountId,
+    block::BlockNumber,
+    note::{Note, NoteDetails, NoteExecutionMode, NoteFile, NoteTag},
+    transaction::TransactionId,
+    utils::Serializable,
+};
+use tokio::sync::mpsc::Sender;
+
+use crate::{EXPLORER_URL, NETWORK_ID};
+
+pub type ResponseSender = Sender<Result<Event, Infallible>>;
+
+/// Sends updates on the minting process to all the clients waiting for a batch of mint requests to
+/// be processed.
+pub struct ClientUpdater {
+    clients: Vec<ResponseSender>,
+}
+
+impl ClientUpdater {
+    /// Creates a new client updater.
+    pub fn new(clients: Vec<ResponseSender>) -> Self {
+        Self { clients }
+    }
+
+    /// Sends an update to all the batch clients.
+    /// Errors when sending through the channel are ignored since the client may have cancelled the
+    /// request.
+    pub async fn send_updates(&self, update: MintUpdate) {
+        let event = update.into_event();
+        for sender in &self.clients {
+            let _ = sender.send(Ok(event.clone())).await;
+        }
+    }
+
+    /// Sends a serialized note to all the batch clients. Each note is sent to the corresponding
+    /// client.
+    /// Errors when sending through the channel are ignored since the client may have cancelled the
+    /// request.
+    pub async fn send_notes(
+        &self,
+        block_number: BlockNumber,
+        notes: &[Note],
+        tx_id: TransactionId,
+    ) {
+        for (note, sender) in notes.iter().zip(&self.clients) {
+            let _ = sender
+                .send(Ok(MintUpdate::Minted(note.clone(), block_number, tx_id).into_event()))
+                .await;
+        }
+    }
+}
+
+/// The different stages of the minting process.
+pub enum MintUpdate {
+    // TODO: add PoW verification event
+    Built,
+    Executed,
+    Proven,
+    Submitted,
+    Minted(Note, BlockNumber, TransactionId),
+}
+
+impl MintUpdate {
+    /// Converts the mint update into an sse event.
+    /// Event types:
+    /// - `MintUpdate::Built`: event type "update"
+    /// - `MintUpdate::Executed`: event type "update"
+    /// - `MintUpdate::Proven`: event type "update"
+    /// - `MintUpdate::Submitted`: event type "update"
+    /// - `MintUpdate::Minted`: event type "note". Contains the note encoded in base64.
+    pub fn into_event(self) -> Event {
+        match self {
+            MintUpdate::Minted(note, block_height, tx_id) => {
```

Comment:
Does this mean that the notes will be sent to the frontend even if it's a public note? If so, it would be nice to make this optional to avoid unnecessary data transfers.